src/Features/Core/Portable/NavigationBar/NavigationBarItems/RoslynNavigationBarItem.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Runtime.Serialization;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.NavigationBar
+{
+    /// <summary>
+    /// Base type of all C#/VB navigation bar items.  Only for use internally to roslyn.
+    /// </summary>
+    [DataContract]
+    internal sealed class RoslynNavigationBarItem
+    {
+        [DataMember(Order = 0)]
+        public readonly RoslynNavigationBarItemKind Kind;
```

Comment:
this is very much a union type.  there is a kind, a set of values always available, and a set of values only aailable depending on the kind.  I really didn't mind coding against this.  But if you hate it, let me know. :)